flink/v1.12/flink/src/main/java/com/netease/arctic/flink/read/source/log/pulsar/LogMsgWithRetractInfo.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic.flink.read.source.log.pulsar;
+
+import com.netease.arctic.flink.read.source.log.kafka.LogKafkaPartitionSplit;
+import com.netease.arctic.log.LogData;
+import org.apache.flink.connector.pulsar.source.reader.message.PulsarMessage;
+import org.apache.pulsar.client.api.MessageId;
+
+public class LogMsgWithRetractInfo<T> extends PulsarMessage<T> {
+
+  /**
+   * Denote reader is in retracting read mode.
+   * In this mode, data would be read in reverse order and opposite RowKind.
+   */
+  private final boolean retracting;
+  /**
+   * @see LogKafkaPartitionSplit#retractStopOffset
+   */
+  private final Long retractStoppingOffset;
+  /**
+   * @see LogKafkaPartitionSplit#revertStartOffset
+   */
+  private final Long revertStartingOffset;
+  /**
+   * @see LogKafkaPartitionSplit#retractingEpicNo
+   */
+  private final Long retractingEpicNo;
+  /**
+   * Data in source, whose {@link LogData#getActualValue()} is the value in log-store.
+   */
+  private final LogData<T> logData;
+  private final T valueToBeSent;
```

Comment:
What does this variable do? 

What's the different with `logData.getActualValue()`?

Reply:
It‘s the value need to be sent to downstream. logData.getActualVaule() is the value in log-store.